### PR TITLE
[opensprinkler] Allow https URLs and basic Auth

### LINKED ISSUE #3540: [opensprinkler] Feature-Request: HTTPS and HTTP basic authentication
Currently it is not possible to use OpenSprinkler in a secure manner with HTTPS or even HTTP basic authentication  (e.g. behind NGINX Reverse Proxy).

As far as I could see the usage of HTTP is fix by usage of a constant named HTTP_REQUEST_URL_PREFIX

https://github.com/openhab/openhab2-addons/blob/fc3988a9054f19d7fac88a0c0ac1c059f7245750/addons/binding/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApiFactory.java#L41

My current knowledge of the whole openHAB development is not sufficient to tackle this myself. So I submit this ticket in hope someone with more superpowers will take care of this.

Regards
Ingo

### PR DESCRIPTION
**Pull request by FlorianSW**
With this PR, the opensprinkler binding will be able to handle https URLs (by adding https:// in front of the hostname), as well as basic authentication.

Fixes #3540

### CODE REVIEW

**Comment by FlorianSW** [comment]
Rebased and resolved the merge conflicts :)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java [hunk 1]
```diff
@@ -82,7 +98,7 @@ public boolean isManualModeEnabled() {
     @Override
     public void enterManualMode() throws CommunicationApiException {
         try {
-            Http.sendHttpGet(getBaseUrl(), getRequestRequiredOptions() + "&" + CMD_ENABLE_MANUAL_MODE);
+            http.sendHttpGet(getBaseUrl(), getRequestRequiredOptions() + "&" + CMD_ENABLE_MANUAL_MODE);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Is there a special reason why this binding is not using the shared httpClient? I guess it should be possible to refactor that without too much effort. The idea is that a `HttpClientFactory` is passedas reference to the handler factory, and the shared client https://github.com/openhab/openhab2-addons/blob/1829ddbed0ed2ccfe2ea6dedf2a4a0b74e8979bd/bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/AVMFritzHandlerFactory.java#L131 is passed to the handler. The handler uses it like https://github.com/openhab/openhab2-addons/blob/1829ddbed0ed2ccfe2ea6dedf2a4a0b74e8979bd/bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzAhaWebInterface.java#L248-L259

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -29,6 +29,14 @@
 				<description>Specifies the refresh interval in seconds.</description>
 				<default>60</default>
 			</parameter>
+			<parameter name="basicUsername" type="text">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/ESH-INF/thing/thing-types.xml]: Is this an advanced configuration option? There are no clear rules but I think if less than 20% of the users need it, you should mark it as `<advanced>true</advanced>`. There is no difference for textual configuration but UIs can decide to hide it until full options are requested.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApiFactory.java [hunk 3]
```diff
@@ -35,49 +31,39 @@
...
+            return new OpenSprinklerHttpApiV210(hostname, port, password, basicUsername, basicPassword);
         } else if (version >= 213) {
-            return new OpenSprinklerHttpApiV213(hostname, port, password);
+            return new OpenSprinklerHttpApiV213(hostname, port, password, basicUsername, basicPassword);
         } else {
             /* Need to make sure we have an older OpenSprinkler device by checking the first station. */
             try {
-                returnContent = Http.sendHttpGet(HTTP_REQUEST_URL_PREFIX + hostname + ":" + port + "/sn0", null);
+                lowestSupportedApi.isStationOpen(0);
             } catch (Exception exp) {
```
[the remark this answers is not in the document]
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApiFactory.java]: Just for the record: I'm not constantly catching "Exception" just to make you pointing me to them, I'm not looking at this when I change code, which I will try to do differently in the future! :D

For now: Done! :P

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApiFactory.java [hunk 4]
```diff
@@ -35,49 +31,39 @@
...
+            String basicPassword) throws CommunicationApiException, GeneralApiException {
         int version = -1;
 
+        OpenSprinklerApi lowestSupportedApi = new OpenSprinklerHttpApiV100(hostname, port, password, basicUsername,
+                basicPassword);
         try {
-            returnContent = Http.sendHttpGet(HTTP_REQUEST_URL_PREFIX + hostname + ":" + port + "/" + CMD_OPTIONS_INFO,
-                    null);
+            version = lowestSupportedApi.getFirmwareVersion();
         } catch (Exception exp) {
```
[the remark this answers is not in the document]
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApiFactory.java]: Done!

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java [hunk 5]
```diff
@@ -66,7 +67,7 @@
      * @return True if the station is open, false if it is closed or cannot determine.
      * @throws Exception
      */
-    public abstract boolean isStationOpen(int station) throws Exception;
+    public abstract boolean isStationOpen(int station) throws GeneralApiException, CommunicationApiException;
```
[the remark this answers is not in the document]
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java]: *thumbsup*

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 2 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/ESH-INF/thing/thing-types.xml]: This is absolutely an "advanced" feature, you need to already have a deeper knowledge about this stuff in order to setup the reverse proxy itself. I'll add the advanced thing :)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java (hunk 1 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Sounds legit, I don't see any reason why the binding needs it's own http client :)

Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: The changes for that were a bit bigger, as I made the "pass anything around anything" stuff a bit less annoying. Would like to get your opinion about the approach as well! :)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerGpioApi.java [hunk 6]
```diff
@@ -27,8 +28,9 @@
  * that the OpenSprinkler PI device is using.
  *
  * @author Jonathan Giles, Chris Graham - Initial contribution
+ * @author Florian Schmidt - Refactor class visibility
  */
-public class OpenSprinklerGpioApi implements OpenSprinklerApi {
+class OpenSprinklerGpioApi implements OpenSprinklerApi {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerGpioApi.java]: I'm not against that, but why did you change the visibility? The internal packages are not exported from the OSGi-bundle anyway.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java [hunk 7]
```diff
@@ -242,4 +261,91 @@ private JcResponse statusInfo() throws CommunicationApiException {
...
+            try {
+                response = withGeneralProperties(httpClient.newRequest(url)).method(HttpMethod.POST)
+                        .content(new StringContentProvider(urlParameters)).send();
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                throw new CommunicationApiException("Request to OpenSprinkler device failed: " + e.getMessage());
+            }
+
+            if (response.getStatus() != HTTP_OK_CODE) {
+                throw new CommunicationApiException(
+                        "Error sending HTTP POST request to " + url + ". Got responce code: " + response.getStatus());
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]:
```suggestion
                        "Error sending HTTP POST request to " + url + ". Got response code: " + response.getStatus());
```

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java [hunk 8]
```diff
@@ -242,4 +261,91 @@ private JcResponse statusInfo() throws CommunicationApiException {
         public List<List<Integer>> ps;
         public int rs;
     }
+
+    /**
+     * This class contains static methods for communicating HTTP GET
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]:
```suggestion
     * This class contains helper methods for communicating HTTP GET
```

Can you move that to a own class-file?`I assume you use that to simplify the communication (instead of calling the jetty methods directly), right?

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV210.java [hunk 9]
```diff
@@ -58,7 +62,7 @@ public boolean isStationOpen(int station) throws Exception {
                     + " but station " + station + " was requested for a status update.");
         }
         try {
-            returnContent = Http.sendHttpGet(getBaseUrl() + CMD_STATION_INFO, getRequestRequiredOptions());
+            returnContent = http.sendHttpGet(getBaseUrl() + CMD_STATION_INFO, getRequestRequiredOptions());
         } catch (Exception exp) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV210.java]: I know it's not your fault, but can you have a look at the try/catch-blocks here?

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerGpioApi.java (hunk 6 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerGpioApi.java]: Yes, not to other bundles, however, this bundle _could_ use the class directly. However, the implementation of the `OpenSprinklerApi` are internal implementation details of the api package. An outside class should not even know that there're different implementations. In my opinion, it makes sense, to enforce this encapsulation on the compiler level as well. That's why I made these classes package private, the factory (which is now even easier to use and change, as it's an OSGi component) and the interface are the only information an outside member of the bundle needs to know and see.

I hope that I explained what I mean in an understandable way :D

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java (hunk 7 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Done! :)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java (hunk 8 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Yes, that's the intention. I'm not sure, if it makes sense to move it to it's own class file. This would expose the Http class to other users of, at least, the package, however, it is meant to be strongly tight to the http api implementation. That's why it's an inner class of the Http Api implementation. Does this make sense? :)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV210.java (hunk 9 above)
Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV210.java]: Absolutely!

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java (hunk 8 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Ok. But I would prefer a better name. Maybe HttpHelper or OpenSprinklerHttp or something which sounds more specialized than Http.

Comment (FlorianSW) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV100.java]: Ok, I choosed "HttpRequestSender", as this describes what it is doing :)
